### PR TITLE
Add left aligned cache support.

### PR DESCRIPTION
**Pull request by wang2yn84**

### CODE REVIEW

File: jetstream_pt/engine.py [hunk 1]
```diff
@@ -545,10 +556,13 @@ def generate(
   ) -> tuple[DecodeState, engine_api.ResultTokens]:
     # seq_len = padded_tokens.shape[0]
     pos = decode_state.current_position
-    input_indexes = jnp.full((1,), pos)
-
-    # fill mask first
-    mask = decode_state.mask.at[:, decode_state.current_position].set(0)
+    if self.env.ring_buffer:
+      input_indexes = jnp.full((1,), pos)
```
Comment (FanhaiLu1) [inline jetstream_pt/engine.py:560]: What if we change current_position to [batch_size, 1], can we use same logic do mask for both ring_buffer and onn_ring_buffer?

File: jetstream_pt/engine.py [hunk 2]
```diff
@@ -570,7 +584,19 @@ def generate(
...
     next_token = self._sampling(logits, self.env.batch_size)
-    lens = decode_state.lens + 1
+    if self.env.ring_buffer:
+      input_pos = decode_state.input_pos + 1
+      lens = decode_state.lens + 1
+    else:
+      input_pos = jnp.where(
+          decode_state.input_pos == 0,
+          0,
+          decode_state.input_pos + 1 % self.env.cache_len,
```
Comment (FanhaiLu1) [inline jetstream_pt/engine.py:594]: In non ring buffer case, can input_pos be larger than cache len?

File: jetstream_pt/engine.py [hunk 3]
```diff
@@ -570,7 +584,19 @@ def generate(
...
+      )
+      lens = jnp.where(
+          decode_state.lens == 0, 0, decode_state.lens + 1 % self.env.cache_len
+      )
```
Comment (FanhaiLu1) [inline jetstream_pt/engine.py:598]: If we delete the lens, any impact in non ring buffer case? My understanding there is 0 impact, correct me if i'm wrong.

File: jetstream_pt/engine.py (hunk 1 above)
Comment (FanhaiLu1) [reply jetstream_pt/engine.py:560]: Yes. I mean if we change current_position to [batch_size, 1], different slot can have different the current_position.  For non ring buffer case, the current_position should be same as input_pos.

File: jetstream_pt/engine.py (hunk 2 above)
Comment (FanhaiLu1) [reply jetstream_pt/engine.py:594]: If no, I feel we don't need do % since it never reach the cache len.

File: jetstream_pt/engine.py (hunk 3 above)
Comment (FanhaiLu1) [reply jetstream_pt/engine.py:598]: sounds good, thanks for more insights.

File: jetstream_pt/engine.py (hunk 2 above)
Comment (wang2yn84) [reply jetstream_pt/engine.py:594]: We don't have control for this. Generate() will keep running if no new prefill results are inserted.

Comment (FanhaiLu1) [reply jetstream_pt/engine.py:594]: Thanks for sharing the details!

**Review by FanhaiLu1 — APPROVED** [review]

File: jetstream_pt/engine.py (hunk 1 above)
Comment (wang2yn84) [reply jetstream_pt/engine.py:560]: It will cause performance regression. Please check jax_experiments.py/test7, inserting with batching + position array takes much longer, like x4~x5

**Review by qihqi — APPROVED** [review]
